bite-size-notes/Week3-c_TrajectoryOptimization_Devansh.md

```diff
@@ -0,0 +1,133 @@
+
+**CBF:**
+A Control Barrier Function (CBF) is a function used to design controllers that ensure safety of a dynamical system. It enforces constraints on the system's state to avoid unsafe regions of operation, such as collision or violation of operational limits.
+
+## Usage in Trajectory Optimization
+
+In trajectory optimization, CLFs and CBFs play crucial roles in ensuring both stability and safety of the system's trajectory:
+
+1. **Stability (CLF):** CLFs are used to design controllers that stabilize the system along the desired trajectory. By ensuring that the CLF decreases along the trajectory, the controller can drive the system towards the desired state while maintaining stability.
+
+2. **Safety (CBF):** CBFs are used to enforce safety constraints during trajectory optimization. These constraints could include avoiding collisions with obstacles, staying within operational limits (e.g., velocity, acceleration), or respecting physical constraints of the system. By incorporating CBFs into the optimization problem, the trajectory planner can generate feasible trajectories that guarantee safety.
+
+## Combining CLF & CBF: 
+
+CLF's usually is a function to move towards a target as stated above whereas CBF's is a function which confines the system to be in a defined set. 
+
+Here, 
+
+$$\min_{u} \lVert \mathbf{u^des - u} \rVert_2^2$$
+
+$$\text{subject to} \quad u \in U $$
+
+$$\text{CLF} \quad \nabla V(x)^T f(x, u) <= 0 or -\alpha V(x)$$ where $-\alpha V(x)$ has to be positive only.
+
+$$\text{CBF} \quad \nabla V(x)^T f(x, u) <= -\beta b(x)$$
+
+Sometimes, there could be conditions when the regions of CLF constraint and CBF constraint do not overlap each other. This is when one can encounter infeasible problem. 
+In such a case, we have to provide a slack or let loose on a constraint, so either of the one (CLF constraint or CBF constraint) can come together and form a feasible region. 
+
+
+For such a case: 
+
+$$\min_{u} \lVert \mathbf{u^des - u} \rVert_2^2 + \gamma \epsilon^2$$ where $\gamma$ is the weighting and $\epsilon$ is the slack variable.
+
+$$\text{subject to} \quad u \in U $$
+
+$$\text{CLF} \quad \nabla V(x)^T f(x, u) <= \epsilon$$ where this is Linear in u and \epsilon.
+
+$$\text{CBF} \quad \nabla V(x)^T f(x, u) <= -\alpha b(x)$$ where this is linear in u.
+
+And, $\epsilon >= 0$ , hence Linear in $\epsilon$ 
+
+## Problem Formulation
+The goal of trajectory optimization is to find the trajectory \( x(t) \) that minimizes (or maximizes) a certain cost function \( J \), subject to constraints. Mathematically, it can be formulated as:
+
+### Discrete Time Dynamics
+
+$$\min_{x_(0)...x_k+1, u_(0)...u_(k)} J_termial(x_(k+1)) + \sum_{k=0}^K J(x_(k), u_(k), k)$$
+
+$$\text{subject to} \quad x_{k+1} = f(x_(k), u_(k), k), \quad k=0,...K$$
+
+$$ \quad \quad \quad x_k \in X, \quad u_(k) \in U_(k), \quad x_(0) = x_current or u_(k) \in U_(k)(x_k)$$ 
+
```

Comment:
Some of the Latex formatting could use some tidying up. In the last constraint, I think "u_(k) \in U_(k)(x_k)" is repeated.

Reply:
No, it s not repeated, one states the feasible control region and other is a function of state x. 
I have defined them properly.